content/blog/2019-10-23-sccp.md

```diff
@@ -0,0 +1,263 @@
+their uses with constants or removing dead code. As such, the live ranges of SSA
+variables that come from the same original variable will never interfere with
+each other and can simply be renamed back to the original variable name. If this
+SSA conversion were to be used for other optimizations that do not share this
+property with constant propagation, the conversion back would need to be
+modified.
+
+### Constant Propagation
+
+The primary component of the optimization is the constant propagation analysis
+itself. I implemented it according to the worklist algorithm described in the
+paper. Doing so was mostly straightforward. However, the paper seems to not
+mention one point in the algorithm where it is necessary to add to the worklist,
+which took a while for me to figure out.[^visit-phi] The output of the analysis
+is a mapping from variables to elements of a lattice, which can be ⊤ (the
+variable is undefined), ⊥ (the value is unknown), or a value (the variable is a
+constant). Because the program is in SSA form, the analysis only needs one
+lattice element per variable, instead of a lattice element for each variable at
+each program point. The paper aludes to the fact that a constant propagation
+analysis can gain information from control-flow branches. For example, for any
+Bril branch, we can conclude that the condition variable is true on the first
+out-edge and false on the second. I did not implement this, as it would require
+keeping track multiple lattice positions per variable and would significantly
+complicate the analysis.[^branches]
+
+After completing the analysis, the next step is to use its results to actually
+modify the code by replacing computations with constants and removing dead code.
+Any block that is unreachable can simply be removed from the CFG. In fact, doing
+this can create more dead code to remove if any variable defined elsewhere is
+only used in removed blocks. The next step is to replace the expressions in
+definitions of variables that are known to be constant with the constant value
+itself. Similarly, this could also create more dead code. As such, the last step
+is to remove any definitions of variables that have no uses. After this, the
+program is converted back out of SSA form, the CFG is flattened back into a
+single list of instructions, and the program is output in its standard JSON
+form.
+
+## Evaluation
+
+It has proven difficult to evaluate the effectiveness of this optimization in a
+manner that would accurately reflect its utility. Upon running the optimization
+on several test cases, I have observed that, for programs that are written
+directly by humans in Bril, the optimization tends to exhibit one of two
+behaviors: either the program is effectively unchanged, or it is completely
+evaluated leaving only assignments of constants to variables, and print
+statements with those variables. By running these test cases, however, I have
```

Comment:
Hmm; maybe it would be better to use a benchmark suite with actual, real inputs instead of just baked-in constant inputs… maybe this "collapse" to constants could be avoided by moving the computation into a function takes the inputs as arguments?